src/Core/Generator/Sampler/TimeBasedSampler.cs

```diff
@@ -0,0 +1,134 @@
+
+        /// <summary>
+        /// Initializes a new instance of the TimeBasedSampler class.
+        /// </summary>
+        /// <param name="container">The Azure Cosmos DB container from which to retrieve data.</param>
+        /// <param name="groupCount">Optional. The number of subranges (or groups) to divide the time range into. Defaults to 10.</param>
+        /// <param name="numberOfRecordsPerGroup">Optional. The number of records to retrieve from each subrange. Defaults to 10.</param>
+        /// <param name="maxDays">Optional. The maximum number of days in the past from which to consider records. Defaults to 10.</param>
+        /// <param name="logger">The logger to use for logging information.</param>
+        public TimeBasedSampler(Container container, int? groupCount, int? numberOfRecordsPerGroup, int? maxDays, ILogger logger)
+        {
+            this._groupCount = groupCount ?? GROUP_COUNT;
+            this._numberOfRecordsPerGroup = numberOfRecordsPerGroup ?? RECORDS_PER_GROUP;
+            this._maxDays = maxDays ?? MAX_DAYS;
+
+            this._logger = logger;
+
+            this._cosmosExecutor = new CosmosExecutor(container, logger);
+        }
+
+        /// <summary>
+        /// Retrieves sampled data by performing the following steps:
+        /// 1. Obtains the highest and lowest timestamps from the data.
+        /// 2. Divides the entire time range into a specified number of subranges (or groups).
+        /// 3. Fetches the top N records from each subrange, ordered by timestamp.
+        /// </summary>
+        /// <returns>A list of JsonDocument objects representing the sampled data.</returns>
+        public async Task<List<JsonDocument>> GetSampleAsync()
+        {
+            _logger.LogInformation("Sampling Configuration is Count(per group): {0}, Days (records considered for grouping): {1}, Group Count: {2}", _numberOfRecordsPerGroup, _maxDays, _groupCount);
+
+            // Step 1: Get the highest and lowest timestamps
+            (long minTimestamp, long maxTimestamp) = await GetHighestAndLowestTimestampsAsync();
+
+            _logger.LogDebug("Min Timestamp(UTC): {0}, Max Timestamp(UTC): {1}", DateTimeOffset.FromUnixTimeSeconds(minTimestamp).UtcDateTime, DateTimeOffset.FromUnixTimeSeconds(maxTimestamp).UtcDateTime);
+
+            // Step 2 & 3: Divide the range into subranges and get data
+            return await GetDataFromSubranges(minTimestamp, maxTimestamp, _groupCount, _numberOfRecordsPerGroup);
+        }
+
+        /// <summary>
+        /// Fetches the minimum and maximum timestamps from the data within the specified time range.
+        /// </summary>
+        /// <returns>A tuple containing the minimum and maximum timestamps.</returns>
+        private async Task<(long minTimestamp, long maxTimestamp)> GetHighestAndLowestTimestampsAsync()
+        {
+            List<long> maxTimestamp = await this._cosmosExecutor.ExecuteQueryAsync<long>(MAX_TIMESTAMP_QUERY);
```

Comment:
why would this query return a list of values instead of just one value?

Reply:
`ExecuteQueryAsync` is generic function which returns list of data.